src/python/pants/option/options_fingerprinter.py

```diff
@@ -16,7 +16,24 @@
                                        target_option)
 
 
+# TODO: this class's methods could be wrapped in @memoized_method -- we would have to manage the
+# cache size somehow if this is done or it might blow up.
 class Encoder(json.JSONEncoder):
+  def _encode_dict_wrapping_keys(self, d):
+    return {self.encode(k):v for k, v in d.items()}
+
+  def encode(self, o):
+    if isinstance(o, (tuple, set)):
+      o = list(o)
+    elif isinstance(o, dict):
+      o = self._encode_dict_wrapping_keys(o)
+    elif len(o) == 1:
```

Comment:
This seems like premature optimization.

Reply:
Everything is wrapped in a list (check the `fingerprint` method in `options_fingerprinter.py`), and we need to be able to modify the dict (encoding the keys) before it goes to the `super` method, or there will be an error. This was the first way I could think of to do that -- it's clunky and could be improved, but my idea was not that it was an optimization, specifically.

Reply:
I recognize that I can just modify the calling code to not wrap everything in a list, but wasn't sure whether there was a specific reason for that.

Reply:
I'm pretty sure the correct solution is to not wrap everything in a list, and in combination with your other comment, the body of this method could potentially be made much smaller.